core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -645,42 +636,49 @@ class GroupCoordinator(val brokerId: Int,
   }
 
   def onCompleteJoin(group: GroupMetadata) {
+    var delayedStore: Option[DelayedStore] = None
     group synchronized {
-      val failedMembers = group.notYetRejoinedMembers
-      if (group.isEmpty || !failedMembers.isEmpty) {
-        failedMembers.foreach { failedMember =>
-          group.remove(failedMember.memberId)
-          // TODO: cut the socket connection to the client
-        }
-
-        // TODO KAFKA-2720: only remove group in the background thread
-        if (group.isEmpty) {
-          group.transitionTo(Dead)
-          groupManager.removeGroup(group)
-          info("Group %s generation %s is dead and removed".format(group.groupId, group.generationId))
-        }
+      // remove any members who haven't joined the group yet
+      group.notYetRejoinedMembers.foreach { failedMember =>
+        group.remove(failedMember.memberId)
+        // TODO: cut the socket connection to the client
       }
+
       if (!group.is(Dead)) {
         group.initNextGeneration()
-        info("Stabilized group %s generation %s".format(group.groupId, group.generationId))
-
-        // trigger the awaiting join group response callback for all the members after rebalancing
-        for (member <- group.allMemberMetadata) {
-          assert(member.awaitingJoinCallback != null)
-          val joinResult = JoinGroupResult(
-            members=if (member.memberId == group.leaderId) { group.currentMemberMetadata } else { Map.empty },
-            memberId=member.memberId,
-            generationId=group.generationId,
-            subProtocol=group.protocol,
-            leaderId=group.leaderId,
-            errorCode=Errors.NONE.code)
-
-          member.awaitingJoinCallback(joinResult)
-          member.awaitingJoinCallback = null
-          completeAndScheduleNextHeartbeatExpiration(group, member)
+        if (group.is(Empty)) {
+          info(s"Group ${group.groupId} with generation ${group.generationId} is now empty")
+          delayedStore = Some(groupManager.prepareStoreGroup(group, Map.empty, errorCode => {
+            if (errorCode != Errors.NONE.code) {
+              // we failed to persist the empty group. if we don't retry (which is how
```

Comment:
I left a comment here because I'm not sure yet how we should handle it. When we fail to write the state for a generation normally, we transition back to `PreparingRebalance` and force the members to rejoin. This results in sort of a "ghost generation" which never gets written to the log. If we then have a partition emigration before the rebalance completes, then the last successfully written generation will come back to life. I'm not sure if this is a problem, but it seems it may be safe since we never members never see the rebalance complete until it is successfully written. 

By doing nothing here, we have similar behavior. If partition emigration happens before the group is cleaned up, then the last stable generation will reemerge and the coordinator will probably just time out the members (since they should be gone by now). Again, I am not sure this is actually a problem. If one of those members was still around (maybe because of a temporary network partition), it should be safe for it to continue doing work since no new non-empty generations could have successfully completed a rebalance without writing the new generation to the log.